### ISSUE TITLE
Deprecate before_consume

### ISSUE THREAD

**Issue by klippx** [phobos/phobos#84]
The `#before_consume` hook was introduced due to a need in klarna/phobos_db_checkpoint in particular the PhobosDBCheckpoint::Handler. 

We introduced it to decode avro binary into json, since phobos_db_checkpoint assumes payload is always a json and we want to store it in a json column. 

However, I am not too sure it really is desired to have this logic in Phobos. 
- The name is not great (sends Rails vibes)
- Hooks are maybe not considered the best design pattern
- Not sure if it is THAT useful? Other than our own use case...
- Has not obvious behaviour (with respect to the name and understanding of Phobos): 
  - It "massages" the payload that will be used downstream
  - It NEEDS to return the payload

WDYT - what do you guys use `#before_consume` for?

**Comment by michberr** [comment]
@klippx thanks for providing some background on before_consume. We don't use phobos_db_checkpoint, so I'm not that familiar with it. 

We currently use `#before_consume` for the following reasons:
1) It helps modularize the code a bit as our `.around_consume` method was getting pretty bloated. This in itself is not essential. We can easily just add more helper methods to `.around_consume`.
2) We use it to catch payload "pre-processing" errors.  We have logic in `#before_consume` to decrypt, decompress, and decode our messages. In general, we want to stall processing if any of these functions throw an exception, because the exception will likely affect all other messages. Again, we can also do this in `.around_consume`, but handling it in`#before_consume` felt convenient.

**One important note:**
The way that `.around_consume` is currently implemented, it does not yield changes to the payload to `handler.consume`. This is somewhat of a gotcha. We currently have a monkey patch on `#process_message` so that any changes to the payload in `.around_consume` are passed on to `#consume`:
```
def process_message(payload)
  ...
  @listener.handler_class.around_consume(preprocessed_payload, @metadata) do |around_consume_payload|
    handler.consume(around_consume_payload, @metadata)
   end
end
```

**TLDR**
I am agnostic about deprecating `#before_consume`; I can easily use `#/.around_consume` for any of my current use cases. However, the implementation of process message would need to change to allow `#/.around_consume` to yield the updated payload to `#consume`.

**Comment by klippx** [comment]
Slightly off topic:

Interesting monkey patch. I was just thinking about yielding the payload/metadata from `around_consume` yesterday when I was looking at the code. Maybe the flow could in fact be like this:

```ruby
def process_message(payload)
  ...
  handler.before_consume(payload, @metadata) do |before_consume_payload|
    handler.around_consume(before_consume_payload, @metadata) do |around_consume_payload|
      handler.consume(around_consume_payload, @metadata)
    end
  end
end
```

Looks a bit crazy and would be quite a breaking change, but useful nonetheless. I think this could be a valid change request.

**Comment by klippx** [comment]
Back on topic. I realize now that `before_consume` is in fact an `around_around_consume` which is totally weird, but it gave me an idea... So instead of `before_consume` (contrived name isn't it) we could implement _any_ preprocess payload logic as a specialized Handler which extend any other handler already in use. 

For example, implementing an AvroHandler wrapping a PhobosDBCheckpoint::Handler:

```ruby
module AvroHandler
  include PhobosDBCheckpoint::Handler

  def around_consume(payload, metadata)
    decoded_payload = avro.decode(payload)
    super(decoded_payload, metadata)
  end
end
```

This would make our code and Phobos code much cleaner and totally get rid of `before_consume` (not even move it to PhobosDBCheckpoint).

WDYT? I guess you could write a custom Handler each for your decrypt, decompress and decode and then compose them as necessary. I guess the risk is that this would make your code much more folded and spread out, but could on the flip side lead to specialized handlers with specialized tests that are potentially "better" structured even though it's harder to read the code for a beginner? Oh, I digress...

**Comment by michberr** [comment]
Totally! I was going to make the same comment that that particular implementation would turn `#before_consume` into another `#around_consume` method 😆 . 

I like your second idea because it gives the user the option to make their code as simple or as layered as needed. In the basic case, a user does not have to worry about a clunky `#before_consume` method, and can simply utilize `#around_consume`. However, if they are also using PhobosDBCheckpoint or have substantial code that they want to factor out from `#around_consume` then they can add as much composition as they see fit. The module approach can be a recommended pattern in the documentation.

**Comment by klippx** [comment]
I guess a nice way forward could be to do a 2 step rocket:

### Step 1
- Deprecate `before_consume` with a message, in a non-breaking way, keeping functionality and suggesting migration path forward
- Deprecate yielding from `around_consume` without also yielding a payload object with a message, keeping functionality and suggesting migration path forward
- Adding specs for backwards compatibility as you guys nicely did for the changes in #82 

### Step 2
- Remove deprecation messages
- Actually remove before_consume (breaking change)
- Remove backwards compatibility (breaking change) from `around_consume`

I am not sure it is easy or possible, sounds like a fun challenge to anyone? 😀

**Comment by MiralDesai** [comment]
Not sure if this is still up for debate but I'll add how I use `before_consume` having just started using phobos.

We have a base consumer class which only has a` before_consume` method, and it's only purpose is to decode the Avro message. That way the data is already a ruby hash when it's gets to the `consume` method. Keeps things quite tidy, our `consume` method then only handles application logic and dealing with the the payload data. Looks like the same thing can be achieved with the new plans for `around_consume`.

**Comment by dorner** [comment]
Yep, we are using it for the exact same purpose. :) The only thing for me would be to make sure we are yielding both payload and metadata - most importantly the key, as that can also be Avro-encoded.

**Comment by dorner** [comment]
This can be closed with 2.0!
